token/token_test.go

```diff
@@ -0,0 +1,194 @@
+	assert.Equal(t, int64(3600), token.ttl)
+
+	token.ttl = 7200
+	assert.Equal(t, int64(7200), token.ttl)
+}
+
+func TestCopyToken(t *testing.T) {
+	t.Parallel()
+	token := New("username", "password", "rawToken", time.Now(), time.Now(), 3600)
+	copiedToken := copyToken(token)
+
+	assert.Equal(t, token.username, copiedToken.username)
+	assert.Equal(t, token.password, copiedToken.password)
+	assert.Equal(t, token.rawToken, copiedToken.rawToken)
+	assert.Equal(t, token.ttl, copiedToken.ttl)
+	assert.Equal(t, token.expiresOn, copiedToken.expiresOn)
+	assert.Equal(t, token.receivedAt, copiedToken.receivedAt)
+
+	// change the copied token
+	copiedToken.expiresOn = time.Now().Add(-1 * time.Hour)
+	assert.NotEqual(t, token.expiresOn, copiedToken.expiresOn)
+
+	// copy nil
+	copiedToken = copyToken(nil)
+	assert.Nil(t, copiedToken)
+	// copy empty token
+	copiedToken = copyToken(&Token{})
+	assert.NotNil(t, copiedToken)
+	anotherCopy := copiedToken.Copy()
+	anotherCopy.rawToken = "changed"
+	assert.NotEqual(t, copiedToken, anotherCopy)
+}
+
+func TestTokenCompare(t *testing.T) {
+	t.Parallel()
+	// Create two tokens with the same credentials
+	token1 := New("username", "password", "rawToken", time.Now(), time.Now(), 3600)
+	token2 := New("username", "password", "rawToken", time.Now(), time.Now(), 3600)
+	assert.True(t, token1.compareCredentials(token2))
+	assert.True(t, token1.compareRawCredentials(token2))
+	assert.True(t, token1.compareToken(token2))
+
+	// Create two tokens with different credentials and different raw credentials
+	token3 := New("username", "differentPassword", "differentRawToken", time.Now(), time.Now(), 3600)
+	assert.False(t, token1.compareCredentials(token3))
+	assert.False(t, token1.compareRawCredentials(token3))
+	assert.False(t, token1.compareToken(token3))
+
+	// Create token with same credentials but different rawCredentials
+	token4 := New("username", "password", "differentRawToken", time.Now(), time.Now(), 3600)
+	assert.False(t, token1.compareRawCredentials(token4))
+	assert.False(t, token1.compareToken(token4))
+	assert.True(t, token1.compareCredentials(token4))
+}
+
+func TestTokenReceivedAt(t *testing.T) {
+	t.Parallel()
+	// Create a token with a specific receivedAt time
+	receivedAt := time.Now()
+	token := New("username", "password", "rawToken", time.Now(), receivedAt, 3600)
+
+	assert.True(t, token.receivedAt.After(time.Now().Add(-1*time.Hour)))
+	assert.True(t, token.receivedAt.Before(time.Now().Add(1*time.Hour)))
+
+	// Check if the receivedAt time is set correctly
+	assert.Equal(t, receivedAt, token.ReceivedAt())
+
+	tcopiedToken := token.Copy()
+	// Check if the copied token has the same receivedAt time
+	assert.Equal(t, receivedAt, tcopiedToken.ReceivedAt())
+	// Check if the copied token is not the same instance as the original token
+	assert.NotSame(t, token, tcopiedToken)
+	// Check if the copied token is a new instance
+	assert.NotNil(t, tcopiedToken)
+}
+
+func BenchmarkNew(b *testing.B) {
```

Comment:
Most of these benchmarks are on basic getters with very low complexity, I don't know how useful they would be, in reality.